python/lsst/daf/butler/script/removeRuns.py

```diff
@@ -0,0 +1,121 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+from collections import defaultdict
+from dataclasses import dataclass
+from functools import partial
+from typing import cast, Callable, Dict, List, Mapping, Sequence, Tuple
+
+from .. import Butler
+from ..registry import CollectionType, MissingCollectionError
+from ..registry.queries import DatasetQueryResults
+
+
+@dataclass
+class RemoveRunsResult:
+    """Container to return to the cli command; contains the names of runs that
+    will be deleted, and a map of dataset type to how many of that dataset will
+    be deleted. Also contains the callback funciton to execute the remove upon
+    user confirmation.
+    """
+    # the callback function to do the removal
+    onConfirmation: Callable[[], None]
+    # list of the run collections that will be removed
+    runs: Sequence[str]
+    # mapping of dataset type name to how many will be removed.
+    datasets: Mapping[str, int]
+
+
+def _getCollectionInfo(
+    repo: str,
+    collection: str,
+) -> Tuple[List[str], Mapping[str, int]]:
+    """Get the names and types of collections that match the collection
+    string.
+
+    Parameters
+    ----------
+    repo : `str`
+        The URI to the repostiory.
+    collection : `str`
+        The collection string to search for. Same as the `expression`
+        argument to `registry.queryCollections`.
+
+    Returns
+    -------
+    runs, datasets : [List[str], Mapping[str, int]]
+        The runs that will be removed, and the dataset types and how many will
+        be removed.
+    """
+    butler = Butler(repo)
+    try:
+        collectionNames = list(
+            butler.registry.queryCollections(
+                collectionTypes=frozenset((CollectionType.RUN,)),
+                expression=collection,
+                includeChains=False,
+            )
+        )
+    except MissingCollectionError:
+        collectionNames = list()
+    runs = []
+    datasets: Dict[str, int] = defaultdict(int)
+    for collectionName in collectionNames:
+        if butler.registry.getCollectionType(collectionName).name == "RUN":
+            runs.append(collectionName)
+            all_results = butler.registry.queryDatasets(..., collections=collectionName)
+            for r in cast(DatasetQueryResults, all_results).byParentDatasetType():
```

Comment:
@TallJimbo thanks for the code you gave me to get the number of datasets (https://lsstc.slack.com/archives/C2JPT1KB7/p1638824513252900?thread_ts=1638823786.251200&cid=C2JPT1KB7)
I wasn't able to get this past mypy without a cast tho. Is there a better way to do this for the type checker?

Reply:
I think you're basically stuck with a cast or an assert, unless we annotate `queryDatasets` with `typing.overload`: the fact that you passed `...` guarantees a more-derived return type for which this code is valid, but passing a `DatasetType` instance returns a different derived type for which it isn't, and mypy doesn't know about any of that.

Reply:
This is all caused by remote registry not knowing about DatasetQueryResults -- in that case it might be better to do it with an instance check rather than a cast so that RemoteRegistry being used here will break early.